backend/zkppSalt.js

```diff
@@ -0,0 +1,229 @@
+
+// TODO HARDCODED VALUES
+const recordPepper = 'pepper'
+const recordMasterKey = 'masterKey'
+const challengeSecret = 'secret'
+const registrationSecret = 'secret'
+const maxAge = 30
+
+const getZkppSaltRecord = async (contract: string) => {
+  const recordId = blake32Hash(hashStringArray('RID', contract, recordPepper))
+  const record = await sbp('chelonia/db/get', recordId)
+
+  if (record) {
+    const encryptionKey = hashStringArray('REK', contract, recordMasterKey).slice(0, nacl.secretbox.keyLength)
+
+    const recordBuf = Buffer.from(record.replace(/_/g, '/').replace(/-/g, '+'), 'base64')
+    const nonce = recordBuf.slice(0, nacl.secretbox.nonceLength)
+    const recordCiphertext = recordBuf.slice(nacl.secretbox.nonceLength)
+    const recordPlaintext = nacl.secretbox.open(recordCiphertext, nonce, encryptionKey)
+
+    if (!recordPlaintext) {
+      return null
+    }
+
+    const recordString = Buffer.from(recordPlaintext).toString('utf-8')
+
+    try {
+      const recordObj = JSON.parse(recordString)
+
+      if (!Array.isArray(recordObj) || recordObj.length !== 3 || !recordObj.reduce((acc, cv) => acc && typeof cv === 'string', true)) {
+        return null
+      }
+
+      const [hashedPassword, authSalt, contractSalt] = recordObj
+
+      return {
+        hashedPassword,
+        authSalt,
+        contractSalt
+      }
+    } catch {
+      // empty
+    }
+  }
+
+  return null
+}
+
+const setZkppSaltRecord = async (contract: string, hashedPassword: string, authSalt: string, contractSalt: string) => {
+  const recordId = blake32Hash(hashStringArray('RID', contract, recordPepper))
+  const encryptionKey = hashStringArray('REK', contract, recordMasterKey).slice(0, nacl.secretbox.keyLength)
+  const nonce = nacl.randomBytes(nacl.secretbox.nonceLength)
+  const recordPlaintext = JSON.stringify([hashedPassword, authSalt, contractSalt])
+  const recordCiphertext = nacl.secretbox(Buffer.from(recordPlaintext), nonce, encryptionKey)
+  const recordBuf = Buffer.concat([nonce, recordCiphertext])
+  const record = recordBuf.toString('base64').replace(/\//g, '_').replace(/\+/g, '-').replace(/=*$/, '')
+  await sbp('chelonia/db/set', recordId, record)
+}
+
+export const getChallenge = async (contract: string, b: string): Promise<false | {authSalt: string; s: string; sig: string;}> => {
+  const record = await getZkppSaltRecord(contract)
+  if (!record) {
+    return false
+  }
+  const { authSalt } = record
+  const s = randomNonce()
+  const now = (Date.now() / 1000 | 0).toString(16)
+  const sig = [now, Buffer.from(hashStringArray(contract, b, s, now, challengeSecret)).toString('base64').replace(/\//g, '_').replace(/\+/g, '-').replace(/=*$/, '')].join(',')
+
+  return {
+    authSalt,
+    s,
+    sig
+  }
+}
+
+const verifyChallenge = (contract: string, r: string, s: string, userSig: string): boolean => {
+  // Check sig has the right format
+  if (!/^[a-fA-F0-9]{1,11},[a-zA-Z0-9_-]{86}(?:==)?$/.test(userSig)) {
+    return false
```

Comment:
@corrideat Under what circumstances would the signature not have the right format?

If it is due to someone entering the wrong password accidentally, I would understand your argument.

But if it is due to hacking or network errors, it needs to be logged along with IP so that `fail2ban` tools can ban the offenders. Or am I missing something?

Reply:
> Under what circumstances would the signature not have the right format?

Mostly for an incorrect implementation or an attack.

Such an attack is mostly harmless, _unless_ logging the requests. Then it becomes a cheap way of filling up space on the server.

Reply:
Consensus from call: logging this at `console.info` level